DEVDOCS-3406: Revise OAuth and API account articles (#1405)
bigcommerce/dev-docs: docs/api-docs/getting-started/authentication.md

# Authentication

BigCommerce offers a suite of APIs that let you manage store data, sign customers in, make client-side queries for product information, and write apps that integrate third-party services into store operations.

This article provides an overview of the authentication schemes that our APIs use organized by the header or other mechanism they use to authenticate with our servers.

> Regardless of authentication scheme, BigCommerce **API accounts** play a role in every authenticated request to our servers.

Author: i take your point. changed to 
```suggestion
Regardless of the subject API's authentication scheme, BigCommerce **API accounts** play a role in every authenticated request to our servers.

```

> Our API accounts come in a few different flavors to meet the needs of different use cases. For example, app API accounts work well in multi-store contexts, whereas store API accounts are a good choice for front-end applications. **Some endpoints only work with one kind of API account.** For a thorough explanation of the differences, check out the section on [choosing the right kind of API account](/api-docs/getting-started/authentication/rest-api-authentication#choosing-the-right-kind-of-api-account) in our API accounts article.

Author: i'm gonna stick with good choice and I hate `frontend` but noted and changed in forthcoming push.

## Stable tokens

> Most of our APIs use credentials that do not expire based on a time frame. Depending on the type of API account the credentials belong to, they may expire based on user or developer actions. 

Author: changed. elaboration on this is in another article to be published shortly.

<!-- For more on events that expire tokens, see ??forthcoming apps article??. -->

> The following sections describe the two authentication schemes we currently have that use stable tokens. They include example requests and indicate the endpoints to which the schemes apply.

Author: they don't actually show. the `access tokens` section describes, and the other one lists.

changed to

```suggestion
The following sections describe the two authentication schemes we use that rely on stable tokens. They include example requests and list or describe the endpoints to which the schemes apply.

```

I'm gonna resolve this because it's one word. feel free to reopen if the solution rankles you!

### Access tokens

<!-- theme: info -->

> > Legacy API accounts used HTTP basic authentication. They are no longer available to new stores. [Learn more about migrating](/api-docs/getting-started/authentication/rest-api-authentication#migrating-from-legacy-to-oauth).

Author: it's an article section, not an article of its own. 

so I did this:

```suggestion
> Legacy API accounts used HTTP basic authentication. They are no longer available to new stores. For more information, see [API Accounts: Migrating from legacy to OAuth](/api-docs/getting-started/authentication/rest-api-authentication#migrating-from-legacy-to-oauth).

```

The `X-Auth-Token` header uses access tokens to authenticate requests. [Create an OAuth API account](/api-docs/getting-started/authentication/rest-api-authentication#all-oauth-api-accounts) to generate access tokens. Pass the access token as the value of the `X-Auth-Token` header of the request you want to authenticate.

For a request to succeed, the access token's API account must have permission to receive the response. Configure your API account with the minimum set of OAuth scopes that your implementation needs. 

To find the specific OAuth scopes your requests require, consult the root API reference pages for the families of endpoints you plan to use. For example, see the [OAuth scopes for the Email Templates endpoints](/api-reference/store-management/email-templates). We also maintain a [list of all our OAuth scopes](/api-docs/getting-started/authentication/rest-api-authentication#oauth-scopes).

#### X-Auth-Token header example requests

> The following tabs contain examples of how to authenticate requests by passing an access token to the `X-Auth-Token` header.

Author: i'm going to decline this change. LMK if you feel strongly.

# Authentication

BigCommerce offers a suite of APIs that let you manage store data, sign customers in, make client-side queries for product information, and write apps that integrate third-party services into store operations.

> Although each class of APIs configures authentication differently, they are all based on the OAuth2.0 standard.

Reviewer: ```suggestion
Although each class of APIs configures authentication differently, they're all based on the OAuth2.0 standard.

```

## Class I: Store Management REST APIs

BigCommerce's Store Management REST APIs use OAuth API account access tokens to authenticate requests. Create a store management OAuth API account to generate and manage access tokens, and pass the access token in the header of the request you want to authenticate.

For more details, see [Obtaining Store API Credentials](/api-docs/getting-started/authentication/rest-api-authentication#obtaining-store-api-credentials).

```http title="Store management REST authentication header"
X-Auth-Token: {{access_token}}
```

<!-- theme: info -->
> Legacy API accounts used HTTP basic authentication. They are no longer available to new stores. [Learn more about migrating](/api-docs/getting-started/authentication/rest-api-authentication#migrating-from-legacy-to-oauth). 

## Class II: GraphQL API

> GraphQL API queries use JWT-style bearer tokens to authenticate. Consult the REST token generation endpoint that corresponds to your GraphQL use case to determine the required OAuth scope before you create a store management OAuth API account to request JWTs for your GraphQL queries. Pass the bearer token in the header of the query you want to authenticate. 

Reviewer 2: Doesn't have to be mentioned here, but might be worth noting that developers using Stencil do not have to generate tokens, as a token is pre-generated for them in the stencil context.

Author: excellent.

Author: done

For more details, see [GraphQL API Authentication](/api-docs/storefront/graphql/graphql-storefront-api-overview#authentication) and [Obtaining Store API Credentials](/api-docs/getting-started/authentication/rest-api-authentication#obtaining-store-api-credentials).

```http title="GraphQL authentication header"
Authorization: Bearer {{generated_jwt}}
```

> ## Class III: App APIs

Reviewer 2: This section confuses me, I don't think "App APIs" are different from "Store Management APIs".

There are separately some JWT-based APIs like Current Customer API and Customer Login API, but I'm not sure they deserve a callout in a section about authentication of API requests generally.

Author: @bookernath I believe that the Provider APIs must be used through apps that partners create. They can't just be implemented on a store with rando store management API credentials. At least, that's what our narrative docs say. @hatertron3000 -- is this right, or do we need to update the narrative docs?

I will modify to clarify that apps can also use all the Store Mgmt APIs and rework these sections for flow. I'm also not married to the Classes -- we'll see how that shakes out.

JWT-based APIs absolutely deserve their own callout in an overview about authentication. Not only do CC and CL use JWTs, but per [my slack question about client IDs](https://bigcommerce.slack.com/archives/C071LJPGX/p1648564063603979), they also require HTTP headers with the client ID to authenticate.

Author: rewritten.

> Apps add novel ways of using BigCommerce store data and can connect third-party services to merchant stores. Not only can apps use all store management REST APIs. Apps authenticate with access tokens that they generate during installation or in response to some app configuration changes. Create an app OAuth API account to enable your app to request and manage BigCommerce-generated access tokens on behalf of merchant stores. You can change the scope of an app OAuth API account at any time. When you modify the OAuth scope, BigCommerce will invalidate the API account's existing access tokens. To facilitate generating new access tokens, BigCommerce will prompt your users to reauthorize the app so that they can accept the new OAuth scope.

Author: fragment addressed, paragraph split up in forthcoming push.

For more on working with apps, see our [Guide to Building Apps](/api-docs/apps/guide/intro). The sections on [Implementing OAuth](/api-docs/apps/guide/auth) and [Callback Handlers](/api-docs/apps/guide/callbacks) are particularly relevant.

```http title="App authentication header for store management APIs"
X-Auth-Token: {{jwt_access_token}}
```

### Shipping Provider API

The Shipping Provider API connects a shipping service to stores and supports a range of shipping-related actions, such as custom rate table calculations. You must create or work within an app to use the Shipping Provider API. Create an app OAuth API account to use it.

Learn more about working with the [Shipping Provider API](/api-docs/providers/shipping).

```http title="App authentication header for the Shipping Provider API"

> X-Auth-Token: {{jwt_access_token}}

Reviewer 2: Is this actually a JWT? I didn't realize this API used different auth from V2/V3 APIs.

Reviewer 3: It's not, and there's actually not any auth headers that BC passes to the shipping provider in rate requests (or requests that an app sends to BC to register itself as a carrier).

DEVDOCS-3621 is open to advise devs on how to authenticate requests from BC to the shipping provider.

Author: @bookernath I guess not --- @hatertron3000's word is good enough for me. 

cool. I'll take a look at 3621.

Author: created DEVDOCS-3757 to update these docs with a link to auth info that will come out of 3621, thanks @hatertron3000

> The Shipping Provider API connects a shipping service to stores and supports a range of shipping-related actions, such as custom rate table calculations. You must create or work within an app to use the Shipping Provider API. Create an app OAuth API account to use it.

Author: elected to keep

```

### Tax Provider API

The Tax Provider API connects tax calculation and filing services to stores. You must create or work within an app to use the Tax Provider API. Create an app OAuth API account to use it.

Learn more about working with the [Tax Provider API](/api-docs/providers/tax).

```http title="App authentication header for the Tax Provider API"

> X-Auth-Token: {{jwt_access_token}}

Reviewer 2: Same question here - does the Tax Provider API really use different auth?

Author: sounds like not.

```

## Class IV: Storefront APIs
### Customer Login API

The Customer Login API facilitates alternative sign-in methods by letting your app generate JWTs that authenticate customers to BigCommerce's servers. Create an app OAuth API account to use the Customer Login API for single sign-on.

For OAuth scope and implementation details, consult [Customer Login API](/api-docs/storefront/customer-login-api) and [SSO API Reference](/api-reference/storefront/customer-login-sso).

### Current Customer API

The Current Customer API enables your app or script to request BigCommerce-generated JWTs that will confirm a customer's identity and logged-in status.
Create an app OAuth API account to use the Current Customer API. 

For OAuth scope and implementation details, consult [Current Customer API](/api-docs/storefront/current-customer-api) and [Get Current Customer](/api-reference/storefront/current-customers/current-customers/getcurrentcustomer).

### Unauthenticated Storefront API endpoints

> The Storefront API allows you to make client-side requests for carts, checkouts, and orders using JavaScript or an alternative language that compiles to run in the browser. It is a convenience collection of operations that affect one customer at a time. These endpoints are unauthenticated and may be low-risk for your store's use case. You can perform authenticated versions of the same operations using GraphQL or the Store Management REST APIs.

Reviewer: ```suggestion
The Storefront API allows you to make client-side requests for carts, checkouts, and orders using JavaScript or an alternative language that compiles to run in the browser. It's a convenience collection of operations that affect one customer at a time. These endpoints are unauthenticated and may be low-risk for your store's use case. You can perform authenticated versions of the same operations using GraphQL or the Store Management REST APIs.

```

## Resources

### Related articles
* [Creating API Accounts](/api-docs/getting-started/authentication/rest-api-authentication)
* [OAuth Scopes](/api-docs/getting-started/authentication/rest-api-authentication#oauth-scopes)

> ### Outside reading

Author: @bookernath -- noted. from an app developer POV, don't certain events like a change in merchant email address or change in the app's token scope expire the token & trigger a new grant code flow the next time the merchant loads the app?
